minoc/src/main/java/org/opencadc/minoc/ArtifactAction.java

```diff
@@ -309,73 +320,75 @@ private void initStorageSite(URI resourceID) {
         }
         log.info("initStorageSite: " + resourceID + " " + name);
         SELF_RESOURCE_ID = resourceID;
     }
     
-    public void checkReadPermission() throws AccessControlException, ResourceNotFoundException, TransientException {
-        PermissionsClient pc = null;
+    public void checkReadPermission()
+        throws AccessControlException, ResourceNotFoundException,
+               TransientException {
+
+        List<Group> userGroups;
+        try {
+            userGroups = getUsersGroups();
+        } catch (PrivilegedActionException e) {
+            throw new IllegalStateException("Error getting user groups", e);
+        }
 
-        /*
         // TODO: optimize with threads
-        for (String readService : readGrantServices) {
-            try {
-                URI serviceID = new URI(readService);
-                // TODO: add this argument to PermissionsClient constructor
-                pc = new PermissionsClient(serviceID);
-                
-                // uncomment when ready
-                pc = new PermissionsClient();
-                ReadGrant grant = pc.getReadGrant(artifactURI);
-                if (grant.isAnonymousAccess()) {
-                    log.debug("anonymous read access granted");
-                    return;
-                }
-                if (grant.getGroups().size() > 0) {
-                    // TODO: check group membership
-                    return;
-                }
+        for (URI readService : readGrantServices) {
+            PermissionsClient pc = new PermissionsClient(readService);
+            ReadGrant grant = pc.getReadGrant(artifactURI);
+            if (grant.isAnonymousAccess()) {
+                log.debug("anonymous read access granted");
                 return;
-            } catch (URISyntaxException e) {
-                throw new IllegalStateException("Invalid read grant service: " + readService);
+            }
+            if (grant.getGroups().size() > 0) {
+                for (GroupURI readGroupUri : grant.getGroups()) {
+                    for (Group userGroup : userGroups) {
+                        if (userGroup.getID() == readGroupUri) {
+                            return;
+                        }
+                    }
+                }
             }
         }
         throw new AccessControlException("read permission denied");
-        */
     }
     
-    public void checkWritePermission() throws AccessControlException, ResourceNotFoundException, TransientException {
-        PermissionsClient pc = null;
-        
+    public void checkWritePermission()
+        throws AccessControlException, ResourceNotFoundException,
+               TransientException {
         // current write auth is simply to be non-anonymous
-        AuthMethod am = AuthenticationUtil.getAuthMethod(AuthenticationUtil.getCurrentSubject());
-        if (am != null && !am.equals(AuthMethod.ANON)) {
+        AuthMethod am = AuthenticationUtil.getAuthMethod(
```

Comment:
remove comment on line 365 since it is no longer true

Reply:
Removed. I'm not checking in ArtifactAction, it should've been part of the next pull request.